### ISSUE TITLE
macro_rules attempts to expand tokens it needs to only pass through.

### ISSUE THREAD

**Issue by eddyb** [rust-lang/rust#28415]
``` rust
macro_rules! expr {
    ($a:expr) => ($a)
}

macro_rules! str_tts {
    ($b:tt) => {
        expr!(stringify!($b))
    }
}

fn main () {
    // This works ("$ c").
    let x = expr!(stringify!($c));
    // This errors with "unknown macro variable `c`".
    let y = str_tts!($c);
    println!("{} {}", x, y);
}
```

[On playpen.](http://is.gd/BYUruW)

It seems that while expanding the `str_tts` invocation, the `$c` token ends up being interpreted as a macro variable, even though it should be packaged in a `tt` non-terminal.

Also to be noted is the fact that `stringify!($c)` produces `"$ c"`, implying there are two tokens, a dollar sign and an identifier, whereas when passed to a `macro_rules` macro, it "fits" in a single `tt`.

Maybe it was a bad idea to ever have a `$` token and a `$var` token type, the `$` token would suffice, although `macro_rules` expansion may become more complicated.

**Comment by Mark-Simulacrum** [comment]
`stringify!($c)`, along with any other macro invocation with `$foo` as an argument, now errors with "error: unknown macro variable `c`", so this can either be closed or renamed/repurposed; a decision as to what we want would be nice.

**Comment by Mark-Simulacrum** [comment]
Not sure what I was talking about before, but today the following error is reported. I believe this indicates that `$c` is now two tokens, so accepting a single tt doesn't work. @jseyfried Can you confirm that interpretation as correct?

```
error: no rules expected the token `c`
  --> test.rs:15:23
   |
15 |     let y = str_tts!($c);
   |             ----------^- in this macro invocation
```

**Comment by eddyb** [comment]
I believe this has been fixed by one of @jseyfried's many token/macro refactors.
(Feel free to wait for proper confirmation, of course)

**Comment by jseyfried** [comment]
@Mark-Simulacrum  Yeah, that interpretation is correct -- that changed was caused by fixing #39403 in #39419.
